lib/OpenQA/Schema.pm

```diff
@@ -14,40 +14,107 @@
 
 package OpenQA::Schema;
-use base qw/DBIx::Class::Schema/;
+use parent qw/DBIx::Class::Schema/;
 use strict;
+use warnings;
+
+use DBIx::Class::DeploymentHandler;
 use Config::IniFiles;
-use IO::Dir;
-use SQL::SplitStatement;
-use Fcntl ':mode';
-use FindBin qw($Bin);
+use Cwd qw/abs_path/;
+use Try::Tiny;
+use FindBin qw/$Bin/;
 
 # after bumping the version please look at the instructions in the docs/Contributing.asciidoc file
 # on what scripts should be run and how
-our $VERSION = 36;
+our $VERSION   = 36;
+our @databases = qw/MySQL SQLite PostgreSQL/;
 
 __PACKAGE__->load_namespaces;
 
-
-
-
 sub connect_db {
     my $mode = shift || $ENV{OPENQA_DATABASE} || 'production';
+    my $prepare_mode;
+    if ($mode eq 'prepare_production') {
+        $prepare_mode = 1;
+        $mode         = 'production';
+    }
     CORE::state $schema;
     unless ($schema) {
         my %ini;
         my $cfgpath = $ENV{OPENQA_CONFIG} || "$Bin/../etc/openqa";
         tie %ini, 'Config::IniFiles', (-file => $cfgpath . '/database.ini');
         $schema = __PACKAGE__->connect($ini{$mode});
+
+        if (!$prepare_mode) {
+            deployment_check($schema);
+        }
     }
     return $schema;
 }
 
 sub dsn {
     my $self = shift;
-
     $self->storage->connect_info->[0]->{dsn};
 }
 
+sub deployment_check {
+    my ($schema) = @_;
+    my $dir = $FindBin::Bin;
+    while (abs_path($dir) ne '/') {
+        last if (-d "$dir/dbicdh");
+        $dir = "$dir/..";
+    }
+    $dir = "$dir/dbicdh";
+    die 'Cannot find database schema files' if (!-d $dir);
+
+    my $dh = DBIx::Class::DeploymentHandler->new(
+        {
+            schema              => $schema,
+            script_directory    => $dir,
+            databases           => \@databases,
+            sql_translator_args => {add_drop_table => 0},
+            force_overwrite     => 0
+        });
+    _try_deploy_db($schema, $dh) or _try_upgrade_db($schema, $dh);
+}
+
+sub _try_deploy_db {
+    my ($schema, $dh) = @_;
+    umask 027;
+    my $version;
+    try {
+        $version = $dh->version_storage->database_version;
+    }
+    catch {
+        if ($schema->dsn =~ /:SQLite:dbname=(.*)/) {
+            # speed this up a bit
+            $schema->storage->dbh_do(
+                sub {
+                    my ($storage, $dbh, @args) = @_;
+                    $dbh->do("PRAGMA synchronous = OFF");
+                });
+        }
+        $dh->install;
+    };
+    return !$version;
+}
+
+sub _try_upgrade_db {
+    my ($schema, $dh) = @_;
+    if ($dh->schema_version > $dh->version_storage->database_version) {
+        if ($schema->dsn =~ /:SQLite:dbname=(.*)/) {
+            # Some SQLite update scripts do not work correctly with foreign keys on
+            $schema->storage->dbh_do(
+                sub {
+                    my ($storage, $dbh, @args) = @_;
+                    $dbh->do("PRAGMA foreign_keys = OFF;");
+                });
+        }
```

Comment:
Twice? One ime is `PRAGMA synchonous = off` and second time `PRAGMA foreign_keys`. Or you mean to hide the same bolierplate?